File: leetcodes/23. Merge k Sorted Lists/mergeKSortedLists.java
import java.util.PriorityQueue;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
  public ListNode mergeKLists(ListNode[] lists) {
    //priority queue with sorting function
    //create new list with 0
    //put all nodes from lists to queue
    //while queue is not empty
    //poll node from queue to new list
      //if node has next node -> put it into queue
    //return new list
    //time O(nlogk) k is the number of linked lists, spaceO(n)
    if (lists == null || lists.length == 0) return null;
    PriorityQueue<ListNode> queue = new PriorityQueue<ListNode>( lists.length, (a, b) -> a.val - b.val);
    ListNode newList = new ListNode(0);
    ListNode temp = newList;
    for (ListNode node: lists){
      if (node != null) queue.add(node);
    }
    while (!queue.isEmpty()){
      temp.next = queue.poll();
      temp = temp.next;
      if (temp.next != null) {
          queue.add(temp.next);
      }
    }
    return newList.next;
  }
}

File: leetcodes/23. Merge k Sorted Lists/mergeKSortedLists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
 var mergeKLists = function(lists) {
  //time O(nlogn), spaceO(n) -> put them to an array, sort the array, put the elems into a listi
  let nodes = []
  let head = new ListNode(0)
  let cur = head
  for (let list of lists) {
    while (list){
      nodes.push(list.val)
      list = list.next
    }
  }
  for (let node of nodes.sort((a,b) =>a-b)){
    cur.next = new ListNode(node)
    cur = cur.next
  }
  return head.next
};